ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLPlanExecuteAndReflectAgentRunner.java

```diff
@@ -0,0 +1,513 @@
+        String parentInteractionId = allParams.get(MLAgentExecutor.PARENT_INTERACTION_ID);
+
+        // completedSteps stores the step and its result, hence divide by 2 to find total steps completed
+        // on reaching max iteration, update parent interaction question with last executed step rather than task to allow continue using
+        // memory_id
+        if (stepsExecuted >= maxSteps) {
+            String finalResult = String
+                .format(
+                    "Max Steps Limit Reached. Use memory_id with same task to restart. \n "
+                        + "Last executed step: %s, \n "
+                        + "Last executed step result: %s",
+                    completedSteps.get(completedSteps.size() - 2),
+                    completedSteps.getLast()
+                );
+            saveAndReturnFinalResult(
+                (ConversationIndexMemory) memory,
+                parentInteractionId,
+                finalResult,
+                completedSteps.get(completedSteps.size() - 2),
+                finalListener
+            );
+            return;
+        }
+
+        MLPredictionTaskRequest request = new MLPredictionTaskRequest(
+            llm.getModelId(),
+            RemoteInferenceMLInput
+                .builder()
+                .algorithm(FunctionName.REMOTE)
+                .inputDataset(RemoteInferenceInputDataSet.builder().parameters(allParams).build())
+                .build(),
+            null,
+            allParams.get(TENANT_ID_FIELD)
+        );
+
+        StepListener<MLTaskResponse> planListener = new StepListener<>();
+
+        planListener.whenComplete(llmOutput -> {
+            ModelTensorOutput modelTensorOutput = (ModelTensorOutput) llmOutput.getOutput();
+            Map<String, String> parseLLMOutput = parseLLMOutput(allParams, modelTensorOutput);
+
+            if (parseLLMOutput.get(RESULT_FIELD) != null) {
+                String finalResult = parseLLMOutput.get(RESULT_FIELD);
+                saveAndReturnFinalResult((ConversationIndexMemory) memory, parentInteractionId, finalResult, null, finalListener);
+            } else {
+                // todo: optimize double conversion of steps (string to list to string)
+                List<String> steps = Arrays.stream(parseLLMOutput.get(STEPS_FIELD).split(", ")).toList();
+                addSteps(steps, allParams, STEPS_FIELD);
+
+                String stepToExecute = steps.getFirst();
+                String reActAgentId = allParams.get(REACT_AGENT_ID_FIELD);
+                Map<String, String> reactParams = new HashMap<>();
+                reactParams.put(QUESTION_FIELD, stepToExecute);
+                if (allParams.containsKey(REACT_AGENT_MEMORY_ID_FIELD)) {
+                    reactParams.put(MEMORY_ID_FIELD, allParams.get(REACT_AGENT_MEMORY_ID_FIELD));
+                }
+
+                reactParams.put(SYSTEM_PROMPT_FIELD, DEFAULT_REACT_SYSTEM_PROMPT);
```

Comment:
This line will always hard code system prompt as default value ? How about check if react system prompt exist or not, if yes, use user provided prompt